### PR TITLE
bonding: T5254: Fixed changing ethernet when it is a bond member

### PR DESCRIPTION
**Pull request by aapostoliuk**
## Change Summary

If ethernet interface is a bond memeber:
  1. Allow for changing only specific parameters which are specified in EthernetIf.get_bond_member_allowed_options function.
  2. Added inheritable parameters from bond interface to ethernet interface which are scpecified in 
      BondIf.get_inherit_bond_options. 
      Users can change inheritable options under ethernet interface but in commit it will be copied from bond interface.
  4. All other parameters are denied for changing.

Added migration script. It deletes all denied parameters under ethernet interface if it is a bond member.
## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes)
- [ ] Migration from an old Vyatta component to vyos-1x, please link to related PR inside obsoleted component
- [ ] Other (please describe):

## Related Task(s)

* https://vyos.dev/T5254

## Component(s) name

bonding

## Proposed changes

If ethernet interface is a bond memeber:
  1. Allow for changing only specific parameters which are specified in EthernetIf.get_bond_member_allowed_options function.
  2. Added inheritable parameters from bond interface to ethernet interface which are scpecified in 
      BondIf.get_inherit_bond_options. 
      Users can change inheritable options under ethernet interface but in commit it will be copied from bond interface.
  4. All other parameters are denied for changing.

Added migration script. It deletes all denied parameters under ethernet interface if it is a bond member.

1. Adding interface with denied parameters to the bond interface
  Configuration:
  set interfaces ethernet eth1 vif 10
  Commands
vyos@vyos# set interfaces bonding bond1 member interface eth1
[edit]
vyos@vyos# commit

Can not add interface "eth1" to bond, it has a "vif 10" assigned!

[[interfaces bonding bond1]] failed
Commit failed
2. Adding interface with inheritable parameters to the bond interface
Configuration:
```
set interfaces bonding bond1 mtu '5000'
set interfaces ethernet eth1 mtu '2000'
```
Commands
```
vyos@vyos# set interfaces bonding bond1 member interface eth1
[edit]
vyos@vyos# commit
[edit]
vyos@vyos# sudo ip link
3: eth1: <NO-CARRIER,BROADCAST,MULTICAST,SLAVE,UP> mtu 5000 qdisc fq master bond1 state DOWN mode DEFAULT group default qlen 1000
    link/ether 0c:ba:0b:5b:00:01 brd ff:ff:ff:ff:ff:ff
8: bond1: <NO-CARRIER,BROADCAST,MULTICAST,MASTER,UP> mtu 5000 qdisc noqueue state DOWN mode DEFAULT group default qlen 1000
    link/ether 0c:ba:0b:5b:00:01 brd ff:ff:ff:ff:ff:ff
[edit]
```
3. Changing denied parameters under ethernet interface
Configuration
`set interfaces bonding bond1 member interface 'eth1'`
Commands
```
vyos@vyos# set interfaces ethernet eth1 vif 10
[edit]
vyos@vyos# commit

Cannot configure "vif 10" on interface "eth1". Interface is a bond
member

[[interfaces ethernet eth1]] failed
Commit failed
[edit]
```
4.Changing inheritable parameters under ethernet interface
Configuration:
```
set interfaces bonding bond1 member interface 'eth1'
set interfaces bonding bond1 mtu '5000'
set interfaces ethernet eth1 mtu '2000'
```
Commands:
```
vyos@vyos# set interfaces ethernet eth1 mtu 9000
[edit]
vyos@vyos# commit

WARNING: Cannot apply "mtu" to "9000". Interface "eth1" is a bond
member. Option is inherited from bond "bond1"

[edit]
vyos@vyos# ip link
3: eth1: <NO-CARRIER,BROADCAST,MULTICAST,SLAVE,UP> mtu 5000 qdisc fq master bond1 state DOWN mode DEFAULT group default qlen 1000
    link/ether 0c:ba:0b:5b:00:01 brd ff:ff:ff:ff:ff:ff
8: bond1: <NO-CARRIER,BROADCAST,MULTICAST,MASTER,UP> mtu 5000 qdisc noqueue state DOWN mode DEFAULT group default qlen 1000
    link/ether 0c:ba:0b:5b:00:01 brd ff:ff:ff:ff:ff:ff
```

5. Removing the interface ethernet from the bond interface
Configuration:
```
set interfaces bonding bond1 member interface 'eth1'
set interfaces bonding bond1 mtu '5000'
set interfaces ethernet eth1 mtu '9000'
```
Commands
```
vyos@vyos# delete interfaces bonding bond1 member interface eth1
[edit]
vyos@vyos# commit
[edit]
vyos@vyos# ip link
3: eth1: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 9000 qdisc fq state DOWN mode DEFAULT group default qlen 1000
    link/ether 0c:ba:0b:5b:00:01 brd ff:ff:ff:ff:ff:ff
8: bond1: <NO-CARRIER,BROADCAST,MULTICAST,MASTER,UP> mtu 5000 qdisc noqueue state DOWN mode DEFAULT group default qlen 1000
    link/ether da:02:ee:d4:35:1d brd ff:ff:ff:ff:ff:ff
[edit]
```

## Checklist:

- [x] I have read the [**CONTRIBUTING**](https://github.com/vyos/vyos-1x/blob/current/CONTRIBUTING.md) document
- [x] I have linked this PR to one or more Phabricator Task(s)
- [x] I have run the components [**SMOKETESTS**](https://github.com/vyos/vyos-1x/tree/current/smoketest/scripts/cli) if applicable
- [x] My commit headlines contain a valid Task id
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly

### CODE REVIEW

**Comment by jestabro** [comment]
Note one minor issue that shows up as I'm testing this:
https://github.com/aapostoliuk/vyos-1x/blob/T5254-1-sagitta/src/migration-scripts/interfaces/30-to-31#L38
will fail on an initial boot, since the low-level ConfigTree.list_nodes fails on non-existence of path. This is avoided in migration scripts by exiting early if base path doesn't exist, for example:
https://github.com/vyos/vyos-1x/blob/current/src/migration-scripts/interfaces/29-to-30#L34-L36

**Review by jestabro — CHANGES_REQUESTED** [review]
Cf. comment above.

**Comment by aapostoliuk** [comment]
> Cf. comment above.

Done

File: src/conf_mode/interfaces-bonding.py [hunk 1]
```diff
@@ -65,6 +67,9 @@ def get_config(config=None):
     Retrive CLI config as dictionary. Dictionary can never be empty, as at least the
     interface name will be added or a deleted flag
     """
+    os.environ['vyos_libexec_dir'] = '/usr/libexec/vyos'
+    os.environ['vyos_validators_dir'] = '/usr/libexec/vyos/validators'
```
Comment (jestabro) [inline src/conf_mode/interfaces-bonding.py]: Why is this needed ?

File: src/migration-scripts/interfaces/30-to-31 [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+if len(argv) < 2:
+    print("Must specify file name!")
+    exit(1)
+
+file_name = argv[1]
+with open(file_name, 'r') as f:
+    config_file = f.read()
+    base = ['interfaces', 'bonding']
+ethernet_commands_config = []
```
Comment (jestabro) [inline src/migration-scripts/interfaces/30-to-31]: Unused def `ethernet_commands_config`

**Review by jestabro — APPROVED** [review]

File: src/conf_mode/interfaces-bonding.py (hunk 1 above)
Comment (aapostoliuk) [reply src/conf_mode/interfaces-bonding.py]: These lines are unused. I deleted them.

File: src/migration-scripts/interfaces/30-to-31 (hunk 2 above)
Comment (aapostoliuk) [reply src/migration-scripts/interfaces/30-to-31]: Done

**Review by sever-sever — APPROVED** [review]

**Review by dmbaturin — APPROVED** [review]

**Comment by c-po** [comment]
@MergifyIo backport sagitta

**Pull request merged**